lib/Doctrine/Annotations/DocParser.php

```diff
@@ -710,28 +753,35 @@ private function Annotation()
             // handle a not given attribute or null value
             if (! isset($values[$valueName])) {
                 if ($property->isRequired()) {
-                    throw AnnotationException::requiredError($propertyName, $originalName, $this->context, 'a(n) ' . $type['value']);
+                    throw AnnotationException::requiredError($propertyName, $originalName, $this->context, 'a(n) ' . $type->describe());
                 }
 
                 continue;
             }
 
-            if ($type !== null && $type['type'] === 'array') {
+            if ($type instanceof ArrayType) {
                 // handle the case of a single value
                 if ( ! is_array($values[$valueName])) {
                     $values[$valueName] = [$values[$valueName]];
                 }
 
-                // checks if the attribute has array type declaration, such as "array<string>"
-                if (isset($type['array_type'])) {
-                    foreach ($values[$valueName] as $item) {
-                        if (gettype($item) !== $type['array_type'] && !$item instanceof $type['array_type']) {
-                            throw AnnotationException::attributeTypeError($propertyName, $originalName, $this->context, 'either a(n) '.$type['array_type'].', or an array of '.$type['array_type'].'s', $item);
+                $valueType = $type->getValueType();
+
+                if (! $type->validate($values[$valueName])) {
+                    $firstInvalidValue = (static function (array $values) use ($valueType) {
```

Comment:
This is awful tbh.

Reply:
It's a temporary hack. Later on I think using exception in validate would be a better approach.

Reply:
I don't want temporary workarounds to be merged. Fix problems reviewers point out. Don't do _I'll just do it later..._

Reply:
This code will be dropped completely so I don'zt really care about that tbh. :D And as said, it's covered by test so it won't be forgotten.
Also, I don't think keys were ever validated so it's basically nop in DocParser.

Reply:
> And as said, it's covered by test 

No you didn't. Which test is testing for keys here?

Reply:
Right, there is no test as the keys were not validated at all, as seen above:
```diff
- foreach ($values[$valueName] as $item) {
```
Currently keys stay of mixed type so key violations can't occur. With new parser it may/will change, right now there is nothing to test.